html/libndt7.js

```diff
@@ -0,0 +1,136 @@
+// libndt7 is a NDTv7 client library in JavaScript.
+const libndt7 = (function () {
+  'use strict';
+
+  // events groups all events
+  const events = {
+    // open is the event emitted when the socket is opened. The
+    // object bound to this event is always null.
+    open: 'ndt7.open',
+
+    // close is the event emitted when the socket is closed. The
+    // object bound to this event is always null.
+    close: 'ndt7.close',
+
+    // error is the event emitted when the socket is closed. The
+    // object bound to this event is always null.
+    error: 'ndt7.error',
+
+    // downloadServer is a event emitted periodically during the
+    // NDTv7 download. It represents a measurement performed by the
+    // server and sent to us over the WebSocket channel.
+    downloadServer: 'ndt7.download.server',
+
+    // downloadClient is a event emitted periodically during the
+    // NDTv7 download. It represents a measurement performed by the client.
+    downloadClient: 'ndt7.download.client'
+  }
+
+  return {
+    // version is the client library version.
+    version: 0.2,
+
+    // events exports the events table.
+    events: events,
+
+    // newClient creates a new NDTv7 client with |settings|.
+    newClient: function (settings) {
+      let funcs = {}
+
+      // emit emits the |value| event identified by |key|.
+      const emit = function (key, value) {
+        if (funcs.hasOwnProperty(key)) {
+          funcs[key](value)
+        }
+      }
+
+      // makequery creates the query string from |settings|.
+      const makequery = function (settings) {
+        let query = ''
+        if (settings.download) {
+          if (settings.download.adaptive || settings.download.duration) {
```

Comment:
The specification supports turning off adaptive and choosing the test duration (within limits imposed by the configuration of the server), because I see cases in which that functionality is useful. We need it currently for gathering data to tune the adaptive algorithm. A regulator may want to measure the sustained speed of a network connection (and there may be private `ndt-cloud` deployments allowing for longer measurements).

This client implements the specification, hence it have all these options. You argument seems to be that the client could be simpler, and by reasoning I gather that this depends also on the specification. So, how about this:

1. we change the specification (now or soon) such that the default is adaptive, such that the default client that we believe one may want to write does not need to encode any URL to obtain the default behaviour

2. we document the parameters that one can pass to the URL and, at least in this JavaScript client, we let the client pass us the query string (if any) that they want to provide to the server

This would simplify the client default implementation but will retain the flexibility to choose.

Does it sound like a reasonable follow up?

Reply:
As we add more options, we make it harder and harder to engage in cross-dataset comparisons, and as such this dilutes the overall value of the MLab ndt7 dataset. This dilution worries me, and is the thing I would most like to avoid. Even worse, the removal of options from a publicly deployed client base is almost impossible.

Why can't we compare adaptive ndt7 downloads to ndt downloads using the old protocol?

Reply:
TL;DR I think your points are valid. I have expressed my reasoning in detail below:

> As we add more options, we make it harder and harder to engage in cross-dataset comparisons, and as such this dilutes the overall value of the MLab ndt7 dataset.

Agreed: it would be more complex if we have rows where the speed is computed using the adaptive algorithm and others in which such algorithm is not used.

> This dilution worries me, and is the thing I would most like to avoid.

Yes, I can see how this could be an additional pain point. And also a place where people could make mistakes by joining rows that should probably not be joined together.

> Even worse, the removal of options from a publicly deployed client base is almost impossible.

A simpler client makes the developer inside me happy. The researcher fears that the comparison may be biased in favour of ndt7 by the usage of non standard ports.

> Why can't we compare adaptive ndt7 downloads to ndt downloads using the old protocol?

I guess it depends on the objective. If the objective is to see whether adaptive ndt7 is not worst than classic NDT, then we can perform this comparison. I guess there may be a follow up research question concerning making the adaptive algorithm better, but I understand that doing this with A/B testing on the server side (and clear labelling of the data) is probably smarter in terms of complexity than doing this from the client side.

Reply:
Okay, I've removed all the options.